### DISCUSSION TITLE
Consecutive php-fpm calls return response for wrong request

### DISCUSSION THREAD

**Discussion by jjmontgo** [brefphp/bref#1200] [Support]
Hello there!

I've been testing Bref out, and so far I love it.  But I've run into an issue that is a definite show-stopper.  I'm hoping someone might have some guidance here because it's pretty unusual.

I have a web page that makes 3 XHR requests to the same endpoint with different URL parameters.  On that endpoint, I print out the content of the $_REQUEST array.  Here are the URL parameters of each request, followed by the output from the $_REQUEST array:

== REQUEST 1 ==

Params: 
?id=3442&categoryId=setup

$_REQUEST:
Array
(
    [id] => 3442
    [categoryId] => setup
)

== REQUEST 2 ==

Params:
?id=3442&categoryId=instructions

$_REQUEST:
Array
(
    [id] => 3442
    [categoryId] => safety
)

== REQUEST 3 ==

Params:
?id=3442&categoryId=safety

$_REQUEST:
Array
(
    [id] => 3442
    [categoryId] => safety
)

As you can see, the second request's URL parameters and $_REQUEST array don't match up.  This happens persistently.

When I print out $_SERVER['REQUEST_URI'], I see ?id=3442&categoryId=safety 

But the browser request is /ajax/activity/loadBlocks/?id=3442&categoryId=instructions

I also printed out the content of $_SERVER['LAMBDA_INVOCATION_CONTEXT'], and the last two requests share exactly the same context information.

If I make a direct request to the endpoint in the browser (not through AJAX), I see the results I would expect.  

It's very unusual.  Can anyone shed some light on how this could happen and what could be done about it?

**Comment by mnapoli** [comment]
I would go as far as to say this is close to impossible that this is what's actually happening 😅 

I mean, it is much more likely that you are sending the wrong requests in JS? How could request 2 have the data of request 3 when request 3 is not even sent?

**Comment by jjmontgo** [comment]
Indeed, that's exactly what I would think.

The production application is running in a typical environment with a load balancer and multiple EC2 instances.  I'm using identical code, an identical database, and the only difference here is the bref/lambda execution environment.  If it were something wrong with the JS alone, it would be happening with both the lambda and the EC2 environment.  

Here are screenshots of the requests in the console, with the content of $_GET printed out from the PHP.  The request query is highlighted with the response on the right.

![pb-lambda-1](https://user-images.githubusercontent.com/11598913/169165060-85501961-e18c-40a4-ab92-b673ad7f8e75.png)

![pb-lambda-2](https://user-images.githubusercontent.com/11598913/169165082-9e4de721-a8f3-4f30-b36c-4e00e0121b4a.png)

![pb-lambda-3](https://user-images.githubusercontent.com/11598913/169165090-567b77b5-c6ae-47da-a8b7-9e5aeba65c3c.png)

This only seems to happen when the 3 requests are chronologically close together.  I can put the URLs in the browser address bar and see correct output.

I've tried using both ${bref:layer.php-73-fpm} and ${bref:layer.php-74-fpm}, but the result is the same.  I'm using the "serverless-lift" plugin, so the request/response is also going through CloudFront.

Is there anything I can output that would explain why this is happening?

**Reply by t-richard** [reply]
> so the request/response is also going through CloudFront

This might be the problem, I know Cloudfront can behave weirdly if cache behaviours are configured incorrectly. Cloudfront should set a response header for cache saying `Miss from Cloudfront`, `Hit from Cloudfront` or something similar. Do you have a hit on the incorrect response?

**Reply by jjmontgo** [reply]
And, you're right.

![pb-lambda-cloudfront](https://user-images.githubusercontent.com/11598913/169420504-d441b4ef-6961-489b-9f7d-c48052365fc6.png)

The first request and the last request are cache misses, while the middle is a hit.

So this means it isn't a lambda problem, but a Cloudfront problem, and I'm super relieved about that.

I don't know why the middle request is cached in Cloudfront when the other two aren't, but now I can focus on that.

Thank you so much @t-richard and @mnapoli

**Comment by t-richard** [comment]
I think it's because those are "same second requests" so cloudfront tries to use cache on them even though the cache policy default to a TTL of zero. This almost look like a bug on cloudfront side. 

You could try to include the query string in the cache behavior to see if it changes something?

You could also try to set cache headers explitly in your code to disable caching.

**Reply by jjmontgo** [reply]
Since the live application is also using Cloudfront, I had a look through the settings there.

The default caching behavior is using "Legacy cache settings" under "Cache key and origin requests", where all headers, query strings and cookies are included in the cache key, and origin cache headers are used.  

The Lambda setup's distribution is using a newer system with a custom caching policy with Min TTL 0, Max TTL 31536000 and Default TTL 0.  The distribution, and the policy, were created by the [serverless-lift](https://github.com/getlift/lift) serverless plugin and its [server-side-website](https://github.com/getlift/lift/blob/master/docs/server-side-website.md) construct.

I changed the caching policy from that to the Managed "Caching Disabled" policy, where all TTL values are 0.  Once I did that, the cache hits stopped and the application now works properly.

So I can conclude that the problem is with serverless-lift / server-side-website.  I'll need to get use an alternative setup.

Thanks again!

**Reply by mnapoli** [reply]
@jjmontgo oh that's super interesting, I'd actually love to fix that in Lift. Do you think you could send a pull request?

**Reply by jjmontgo** [reply]
I'd love to, especially since I've never sent a pull request before and I'm not sure how.

[Looking at the repo, I see that you've decided to not use a caching disabled policy...](https://github.com/getlift/lift/blob/a26b434b8f06497d01a00f9599291d92cd40ef07/src/constructs/aws/ServerSideWebsite.ts#L128-L130)

I don't know why the cache hit occurs, but switching to cache disabled is the only work-around I have.

**Reply by mnapoli** [reply]
Ahh, good point. Here is the config for the official "Caching Disabled" policy (https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-cache-policies.html):

<img width="848" alt="image" src="https://user-images.githubusercontent.com/720328/170944250-3792f9ea-9094-4bda-af3e-6b5442a3cdc6.png">

Maybe what we need to do is explicitly add "minimum" and "maximum" TTL to `0` in Lift's config?

**Reply by jjmontgo** [reply]
Yes, that was the first thing I tried:

![ttl-max-zero-error](https://user-images.githubusercontent.com/11598913/171066383-914fb836-807e-4ec7-851e-7fdea622692b.png)

There was only one google search result which matched this error message, and it had to do with an origin request policy.

I had to remove the Authorization header from the cache key settings, and then I could save the policy.

I believe this made the custom policy equivalent to the "Caching Disabled" policy.

**Reply by mnapoli** [reply]
OK, we could actually maybe switch to the "Caching Disabled" official policy then. Someone else opened a similar issue here too: https://github.com/getlift/lift/issues/144

**Reply by mnapoli** [reply]
Update: I dug into that again and could not reproduce it with a "normal" application.

Looking again at your original screenshot, I see your app seems to emit an `expires` header. I think this is what triggers caching by CloudFront.

So my conclusion is: this is expected, if your app emits a caching header, CloudFront will respect that.

I'll be updating https://github.com/getlift/lift/issues/144 as well to consolidate all discussions.

cc @t-richard FYI

**Reply by mnapoli** [reply]
I'm sorry, I was wrong!

Maybe in your situation it was the `expires` header, but it does look like we have a problem. I've posted an update in https://github.com/getlift/lift/issues/144#issuecomment-1179686086 let's continue the discussion there.
